FreeRTOS-Plus/Source/FreeRTOS-IoT-Libraries-LTS-Beta2/c_sdk/platform/freertos/transport/src/mbedtls_freertos.c

```diff
@@ -0,0 +1,541 @@
+    sslContextInit( &( pNetworkContext->sslContext ) );
+
+    mbedtlsError = mbedtls_ssl_config_defaults( &( pNetworkContext->sslContext.config ),
+                                                MBEDTLS_SSL_IS_CLIENT,
+                                                MBEDTLS_SSL_TRANSPORT_STREAM,
+                                                MBEDTLS_SSL_PRESET_DEFAULT );
+
+    if( mbedtlsError != 0 )
+    {
+        LogError( ( "Failed to set default SSL configuration, error %d.", mbedtlsError ) );
+
+        /* Per mbed TLS docs, mbedtls_ssl_config_defaults only fails on memory allocation. */
+        returnStatus = MBEDTLS_INSUFFICIENT_MEMORY;
+    }
+
+    if( returnStatus == MBEDTLS_SUCCESS )
+    {
+        /* Set up the certificate security profile, starting from the default value. */
+        pNetworkContext->sslContext.certProfile = mbedtls_x509_crt_profile_default;
+
+        /* test.mosquitto.org only provides a 1024-bit RSA certificate, which is
+         * not acceptable by the default mbed TLS certificate security profile.
+         * For the purposes of this demo, allow the use of 1024-bit RSA certificates.
+         * This block should be removed otherwise. */
+        if( strncmp( pHostName, "test.mosquitto.org", strlen( pHostName ) ) == 0 )
+        {
+            pNetworkContext->sslContext.certProfile.rsa_min_bitlen = 1024;
+        }
+
+        /* Set SSL authmode and the RNG context. */
+        mbedtls_ssl_conf_authmode( &( pNetworkContext->sslContext.config ),
+                                   MBEDTLS_SSL_VERIFY_REQUIRED );
+        mbedtls_ssl_conf_rng( &( pNetworkContext->sslContext.config ),
+                              mbedtls_ctr_drbg_random,
+                              &ctrDrgbContext );
+        mbedtls_ssl_conf_cert_profile( &( pNetworkContext->sslContext.config ),
+                                       &( pNetworkContext->sslContext.certProfile ) );
+    }
+
+    if( returnStatus == MBEDTLS_SUCCESS )
+    {
+        /* Parse the server root CA certificate into the SSL context. */
+        mbedtlsError = mbedtls_x509_crt_parse( &( pNetworkContext->sslContext.rootCa ),
+                                               ( const unsigned char * ) pNetworkCredentials->pRootCa,
+                                               pNetworkCredentials->rootCaSize );
+
+        if( mbedtlsError != 0 )
+        {
+            LogError( ( "Failed to parse server root CA certificate, error %d.",
+                        mbedtlsError ) );
+
+            returnStatus = MBEDTLS_INVALID_CREDENTIALS;
+        }
+        else
+        {
+            mbedtls_ssl_conf_ca_chain( &( pNetworkContext->sslContext.config ),
+                                       &( pNetworkContext->sslContext.rootCa ),
+                                       NULL );
+        }
+    }
+
+    if( returnStatus == MBEDTLS_SUCCESS )
+    {
+        if( ( pNetworkCredentials->pPrivateKey != NULL ) && ( pNetworkCredentials->pClientCert != NULL ) )
```

Comment:
Should there be a similar NULL check for `pRootCa` in the previous if-block?

Reply:
No. It could be a configAssert in the function. RootCa is must for mutual auth and server auth. client certificate and keys are only needed for mutual auth.